task/oci-copy/0.1/oci-copy.yaml

```diff
@@ -0,0 +1,157 @@
+          echo "declare OCI_ARTIFACT_TYPE=${artifact_type}" >> /var/workdir/vars/$filename
+          echo "declare OCI_ARTIFACT_DIGEST=${artifact_digest}" >> /var/workdir/vars/$filename
+
+          echo "Wrote /var/workdir/vars/$filename with contents:"
+          cat /var/workdir/vars/$filename
+        done
+      workingDir: $(workspaces.source.path)
+
+    - name: oci-copy
+      # I need something with adequate rights to run:
+      # $ skopeo copy docker://quay.io/containers/buildah:v1.35.4 docker://quay.io/redhat-appstudio/buildah:v1.35.4
+      #
+      # For now, use buildah directly from upstream
+      # image: quay.io/containers/buildah:v1.35.4@sha256:f07da1c4c59290dd7ff8c521f5d8f3b0f24985ae933da7381fd5a2c96f887423
+      image: quay.io/containers/buildah:v1.35.4
+      computeResources:
+        limits:
+          memory: 4Gi
+        requests:
+          cpu: 250m
+          memory: 512Mi
+      securityContext:
+        capabilities:
+          add:
+            - SETFCAP
+      script: |
+        set -eu
+        set -o pipefail
+
+        buildah manifest create $IMAGE
+
+        for varfile in /var/workdir/vars/*; do
+          echo "Reading $varfile"
+          source $varfile
+
+          echo "Downloading $OCI_SOURCE to $OCI_FILENAME"
+          curl --silent --show-error --location $OCI_SOURCE -o $OCI_FILENAME
+
+          echo "Confirming that digest of $OCI_FILENAME matches expected $OCI_ARTIFACT_DIGEST"
+          echo "$OCI_ARTIFACT_DIGEST $OCI_FILENAME" | sha256sum --check
+
+          echo "Building manifest of type $OCI_ARTIFACT_TYPE from $OCI_FILENAME"
+          buildah manifest add $IMAGE --artifact --artifact-type $OCI_ARTIFACT_TYPE $OCI_FILENAME
+        done
+
+        echo "Pushing conents to $IMAGE"
+        buildah manifest push \
+            --digestfile $(workspaces.source.path)/image-digest \
+            --authfile $HOME/.docker/config.json \
+            --all \
+            $IMAGE
+
+        cat "$(workspaces.source.path)"/image-digest | tee $(results.IMAGE_DIGEST.path)
+        echo -n "$IMAGE" | tee $(results.IMAGE_URL.path)
+      volumeMounts:
+        - mountPath: /var/lib/containers
+          name: varlibcontainers
+      workingDir: $(workspaces.source.path)
+    - image: quay.io/konflux-ci/yq:latest@sha256:974dea6375ee9df561ffd3baf994db2b61777a71f3bcf0050c5dca91ac9b3430
+      name: sbom-generate
+      script: |
+        cat >sbom-cyclonedx.json <<EOL
+        {
+            "\$schema": "http://cyclonedx.org/schema/bom-1.5.schema.json",
+            "bomFormat": "CycloneDX",
+            "specVersion": "1.5",
+            "version": 1,
+            "components": []
+        }
+        EOL
+
+        for varfile in /var/workdir/vars/*; do
+          echo "Reading $varfile"
+          source $varfile
+
+          purl="pkg:generic/${OCI_FILENAME}?download_url=${OCI_SOURCE}&checksum=sha256:${OCI_ARTIFACT_DIGEST}"
```

Comment:
Including the raw url in the purl can end up making the purl invalid (to a lesser extent, the same goes for the filename) https://github.com/package-url/purl-spec/blob/master/PURL-SPECIFICATION.rst#character-encoding

Both the filename and the url should ideally be percent-encoded. `jq` has a built-in `@uri` filter, though I think that one encodes too much - per the purl-spec, it sounds like `:` and `/` should never be encoded. Python stdlib's `urllib.parse.quote(safe=":/")` would work.

Reply:
Although it doesn't seem too likely to cause issues in practice. The most likely I think would be source urls containing a query part `?`. So it might be fine to leave as is for now

Reply:
Good idea. Will see if I can adapt it.

Reply:
I suppose we can do this inline?

```
echo 'http://konflux-ci.dev/query?this=example' | python3 -c 'import sys; import urllib; print(urllib.parse.quote(sys.stdin.read().strip(), safe=":/"))' 
```

Reply:
Done in 69cd1b4.